File: iOS/old/精通iOS开发/9/Fonts/Fonts/FavoritesList.m
//
//  FavoritesList.m
//  Fonts
//
//  Created by andyron on 16/2/17.
//  Copyright © 2016年 andyron. All rights reserved.
//

#import "FavoritesList.h"

@interface FavoritesList()

@property (strong, nonatomic) NSMutableArray *favorites;

@end
@implementation FavoritesList

+ (instancetype)sharedFavoritesList
{
    static FavoritesList *shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shared = [[self alloc] init];
    });
    return shared;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        NSArray *storedFavorites = [defaults objectForKey:@"favorites"];
        if (storedFavorites) {
            self.favorites = [storedFavorites mutableCopy];
        } else {
            self.favorites = [NSMutableArray array];
        }
    }
    return self;
}

- (void)addFavorite:(id)item
{
    [_favorites insertObject:item atIndex:0];
    [self saveFavorites];
}
- (void)removeFavorite:(id)item
{
    [_favorites removeObject:item];
    [self saveFavorites];
}
- (void)saveFavorites
{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    [defaults setObject:self.favorites forKey:@"favorites"];
    [defaults synchronize];
}
@end

File: iOS/Beginning iPhone Development with Swift 2, 2nd Edition/9/Fonts/Fonts/FavoritesList.swift
//
//  FavoritesList.swift
//  Fonts
//
//  Created by andyron on 16/8/3.
//  Copyright © 2016年 andyron. All rights reserved.
//

import Foundation
import UIKit

class FavoritesList {
    static let sharedFavoritesList = FavoritesList()
    private(set) var favorites:[String]
    
    init(){
        // 读取用户设置，初始化favorites
        let defaults = NSUserDefaults.standardUserDefaults()
        let storedFavorites = defaults.objectForKey("favorites") as? [String]
        favorites = storedFavorites != nil ? storedFavorites! : []
    }
    
    func addFavorites(fontName:String){
        if !favorites.contains(fontName) {
            favorites.append(fontName)
            saveFavorites()
        }
    }
    
    func removeFavorites(fontName:String){
        if let index = favorites.indexOf(fontName) {
            favorites.removeAtIndex(index)
            saveFavorites()
        }
    }
    private func saveFavorites(){
        let defaults = NSUserDefaults.standardUserDefaults()
        defaults.setObject(favorites, forKey: "favorites")
        defaults.synchronize()
    }
}